compliance_checker/cf/cf.py

```diff
@@ -3698,6 +4035,286 @@ def check_cell_measures(self, ds):
 
         return ret_val
 
+    def check_grid_mapping_attr_condition(self, attr, attr_name):
+        """
+        Evaluate a condition (or series of conditions) for a particular
+        attribute. Implementation for CF-1.7.
+
+        :param attr: attribute to teset condition for
+        :param str attr_name: name of the attribute
+        :rtype tuple
+        :return two-tuple of (bool, str)
+        """
+
+        if attr_name == 'geographic_crs_name':
+            return self._evaluate_geographic_crs_name(attr)
+
+        elif attr_name == 'geoid_name': # TODO
+            return self._evaluate_geoid_name(attr)
+
+        elif attr_name == 'geopotential_datum_name': # TODO
+            return self._evaluate_geopotential_datum_name(attr)
+
+        elif attr_name == 'horizontal_datum_name':
+            return self._evaluate_horizontal_datum_name(attr)
+
+        elif attr_name == 'prime_meridian_name':
+            return self._evaluate_prime_meridian_name(attr)
+
+        elif attr_name == 'projected_crs_name':
+            return self._evaluate_projected_crs_name(attr)
+
+        elif attr_name == 'reference_ellipsoid_name':
+            return self._evaluate_reference_ellipsoid_name(attr)
+
+        elif attr_name == 'towgs84': # TODO
+            return self._evaluate_towgs84(attr)
+
+        else: # invoke method from 1.6, as these names are all still valid
+            return super(CF1_7Check, self).check_grid_mapping_attr_condition(attr, attr_name)
+
+    def _check_gmattr_existence_condition_geoid_name_geoptl_datum_name(self, var):
+        """
+        Check to see if both geoid_name and geopotential_datum_name exist as attributes
+        for `var`. They should not.
+
+        :param netCDF4.Variable var
+        :rtype tuple
+        :return two-tuple (bool, str)
+        """
+
+        msg = 'Both geoid_name and geopotential_datum_name cannot exist'
+
+        if ('geoid_name' in var.ncattrs()) and ('geopotential_datum_name' in var.ncattrs()):
+            return (False, msg)
+
+        else:
+            return (True, msg)
+
+    def _check_gmattr_existence_condition_ell_pmerid_hdatum(self, var):
+        """
+        If one of reference_ellipsoid_name, prime_meridian_name, or
+        horizontal_datum_name are defined as grid_mapping attributes,
+        they must all be defined.
+
+        :param netCDF4.Variable var
+        :rtype tuple
+        :return two-tuple (bool, str)
+        """
+
+        msg = ("If any of reference_ellipsoid_name, prime_meridian_name, "
+               "or horizontal_datum_name are defined, all must be defined.")
+
+        if (not all([
+            x in var.ncattrs() for x in [
```

Comment:
Store `var.ncattrs()` and do a set comparison of `{'reference_ellipsoid_name', 'prime_meridian_name', 'horizontal_datum_name'}` to avoid calling ncattrs() each time.

Reply:
See newest commit for changes